### DISCUSSION TITLE
Padding, Alignment

### DISCUSSION THREAD

**Discussion by OverLordGoldDragon** [kymatio/kymatio#716] [Q&A] [answered]
## Background

Different `psi2` produce different number of second-order time scattering coefficients. Padding all per maximum is an option, but compute expensive; ideally we'll pad as much as we need to, but this introduces (and reveals) more considerations:

 1. Different paddings _break alignment_ of frequency rows; a peak at row 32 of 128 will appear at row 4 of 16 at subsampled output, but if we pad to 64, it's now at 2 of 16. Should we **enforce alignment**, and **how**?
 2. Different paddings require different lengths of frequential filters, `sc_freq.psi1_f`. Do we **resample** or **subsample**? ([moved](https://github.com/kymatio/kymatio/discussions/721))

## 1. Alignment

First a problem summary; below's `psi2[4] * U_1_hat` coeffs for an A.M. cosine:

![image](https://user-images.githubusercontent.com/16495490/114300715-b414be00-9ad2-11eb-98ee-0ef60a4de77d.png)

Center is _minimally padded_, right is to the common maximum. Vertical = freq, horizontal = time. These are then convolved by `sc_freq.psi1_f` filters:

![image](https://user-images.githubusercontent.com/16495490/114300800-fb02b380-9ad2-11eb-8896-61212a807303.png)

Map to Fourier, multiply by lowpass for `ifft` convolution later:

![image](https://user-images.githubusercontent.com/16495490/114300815-0d7ced00-9ad3-11eb-966a-f734e9224be6.png)

Subsample:

![image](https://user-images.githubusercontent.com/16495490/114300853-271e3480-9ad3-11eb-8f3e-7dff034e9275.png)

Final output shape is the same. Note, the two look identical. Looking at real part instead of `abs`, however:

![image](https://user-images.githubusercontent.com/16495490/114300884-46b55d00-9ad3-11eb-8c06-38404518facc.png)

Different phase. Finally `ifft()`:

![image](https://user-images.githubusercontent.com/16495490/114300924-782e2880-9ad3-11eb-9b03-7d69e03dc10f.png)

They look same, except aligned differently. Comparing against first image, this is exactly consistent with the peak rows' _relative positions_ - about halfway along vertical for min-padded, and quarter for max-padded.

**Can we force alignment?** Long story short, yes, but neither easily nor without compute expense; I suggest skimming bottom section.

### Should we force alignment?

Compute aside, it makes a significant difference in our output, beyond alignment. All ops before lowpass + subsampling can be done fully equivalently with less padding. But _at_:

![image](https://user-images.githubusercontent.com/16495490/114303547-af0a3b80-9adf-11eb-83f9-2568ffdc7537.png)

Blue = padded `psi_t * psi_f` coeffs, orange = lowpass, red = subsampling points (but _after_ conv). Lowpasses have same discrete time width.

 1. Left, half of the lowpass is drawing from coeff values, nonzero. Right, the lowpass is drawing from almost all zeros.
 2. Left, we skip 1 sample in the nonzero region when subsampling - right, we skip three.

Thus, right's output is based on operating on more _zeros_, whereas left's more on _coefficients_. Or, left's is **more information dense**. This holds regardless of adjustments to width of lowpass. Depending on how we adjust filters (resample / subsample), it's also a difference in imposed statistical priors.

On a higher level, it's as below, where `fn()` is a uniform operator (same # of inputs contribute to each output): left encodes _more information per output sample_, whereas right _distills it_

<img src="https://user-images.githubusercontent.com/16495490/114304215-11187000-9ae3-11eb-8d98-781cfd72062a.png" width="450">

Back to JTFS, the high-level difference output is: less padding amounts to _zooming_ into the frequential axis, dedicating same number of outputs to fewer frequency rows, thus increasing "fineness" of representation.

Thus, we are to decide between 1) maximizing information per coefficient; 2) enforcing consistent cross-wavelet structure. Latter will aid e.g. 3D conv nets in exploiting cross-wavelet dependencies. But overall, I say former is lot more important.

**VERDICT**: convolve directly with less padding, forget alignment.

## Forcing alignment

**Won't work**:

 - Complex rotation / circular shift: the shift factors are _variable_: e.g. row 4 must move to row 2, and 8 to 4. 
 - Shifting padding indices: problem identical to above

One option is to _re-pad_ the output of `Y_2_hat * sc_freq.psi1_f` to maximal, before lowpass + subsample. We still save compute on `Y_2_hat * sc_freq.psi1_f`, but also add it relative to not re-padding.

Question is, can we do anything _without_ padding back to maximum? After digging, the equivalent problem is:

```python
Need `op` such that

subsamp4(DFT(zeropad(x))) == subsamp2(op(DFT(x)))   [1]
or
subsamp4(DFT(zeropad(x))) == subsamp2(DFT(op(x)))   [2]
or
subsamp4(DFT(zeropad(x))) == op(subsamp2(DFT(x)))   [3]

where

zeropad(x)  == pad(x, [0, len(x)])
subsamp2(x) == x.reshape(2, len(x)//2).mean(axis=0); let len(x)==64
            == [(x[0] + x[32])/2, (x[1] + x[33])/2, ...]
subsamp4(x) == [(x[0] + x[16] + x[32] + x[48])/4, (x[1] + ...]
```

Ideally `[3]` exists... but it doesn't. The other two do, but I've only confirmed for enforcing alignment relative to _minimal_ padding, rather than maximal - i.e. let `padlen=32` set peak positions of `padlen=128`. The other way around seems possible since at stages `[1]` and `[2]` we are still in fully invertible territory.

**Comment by lostanlen** [comment]
Thanks for all the sleuthing. There's something i don't understand with your premise:

> Different paddings break alignment of frequency rows; a peak at row 32 of 128 will appear at row 4 of 16 at subsampled output, but if we pad to 64, it's now at 2 of 16. Should we enforce alignment, and how?

Why would you subsample twice as much in the second case?

My version would rather be:

> A peak at row 128 will appear at row 4 of 16 at subsampled output, but if we pad to 64, it's now 4 of 32.

In this way, everything stays aligned. The only remaining question is how many rows to remove at the end during unpadding. (Answer: 16 more rows in the second example than in the first)

**Reply by OverLordGoldDragon** [reply]
@lostanlen We have a grid of options:

```
  1. Pad varyingly,     subsample varyingly, no trim -> same out shape, unaligned
  2. Pad varyingly,     subsample minimally  no trim -> var  out shape, !aligned
  3. Pad same (to max), subsample maximally, no trim -> same out shape, aligned
  4. Pad same (to max), subsample minimally, trim    -> var  out shape, !aligned

pad varyingly = minimally per each `psi2`, 
subsample varyingly = proportionally to pad
subsample max/min = according to max/min padding
"trim" refers to absolute unpadding amounts increasing to compensate for less subsampling
```

1 is my proposed scheme, 2 is yours. 4 is 2 with more compute. 3 loses extra info for small pads - and like any `aligned` output, early `psi2` (low `j2`) coeffs are mostly zero and wasteful. (3 is main branch's current implementation) 

Note I marked 2 `!aligned` - because it's not actually aligned across all slices as described in my post. That is, 1/5th row isn't for 1/5th frequency, but for 1/5th of `current_pad/max_pad` of frequency. Or, it's aligned linearly with freq axis (same absolute row position = same freq).

So it's actually same "alignment" as mine in information sense, we're still zooming on frequency axis rather than enforcing `row0 = min freq`, `row63 = max freq`, only difference is in output shape - mine being more compressive, but permitting concatenation. Mine also has a unique disadvantage: output is multirate, or of different physical scales.

> first, make it correct (and energy-preserving)

1 is 2 at lower sampling rate without aliasing, energy is preserved.

> partial workaround to this: only compute psi2 above a certain time scale

Much rather not lose info, also adds complexity

<hr>

Let me know if you disagree anywhere

**Reply by OverLordGoldDragon** [reply]
> 1 is 2 at lower sampling rate without aliasing, energy is preserved

Per this, 2 corresponds to a variable `oversampling` along frequency.

**Reply by lostanlen** [reply]
thanks for the report.
i think 3 should be the implementation for `out_type=="array"` and 2 the implementation for `out_type=="list"`

> Or, it's aligned linearly with freq axis (same absolute row position = same freq).

i would call that an alignment :)
it's aligned in the sense of jagged arrays

> mine being more compressive, but permitting concatenation. 

alignment is necessary to permit concatenation. If the output sizes are the same but the contents are mismatched, i don't see the point in concatenating.

> Mine also has a unique disadvantage: output is multirate, or of different physical scales.

sounds like a deal-breaker to me. If your axis doesn't have a physical interpretation, what's the point of it?


Tying together the subsampling and the padding and training a CNN on that representation is an interesting (if risky) research idea but i really don't think it should be the default in Kymatio. it will be impossible to document clearly and it will confuse every newcomer.

**Reply by OverLordGoldDragon** [reply]
> alignment is necessary to permit concatenation

@lostanlen 1 is aligned in same sense as 2, on per-`psi2` basis

> doesn't have a physical interpretation

It does: scale doubles across increasing `j2`. Per `j2` is consistent. 

> impossible to document clearly

```
opt: if True, will subsample every joint coefficient along frequency to same output shape. 
Output is then multirate, with earliest coefficients (low j2) having finest frequential resolution 
(most frequency rows per time scattering rows), then halved with increasing `j2`.
```
> don't think it should be the default in Kymatio

I don't mind, as long as it's an option.

<hr>

For classification, I'd like to stress the following: 

> 1 is 2 at lower sampling rate without aliasing, energy is preserved
Thus, 2 corresponds to a variable oversampling along frequency.

So _per-slice_, 2 is just upsampled 1. _There is no additional information_.

**Reply by lostanlen** [reply]
OK, i do i extract the response of JTFS in the octave 400-800 Hz ?

**Reply by lostanlen** [reply]
> opt: if True, will subsample every joint coefficient along frequency to same output shape. 
Output is then multirate, with earliest coefficients (low j2) having finest frequential resolution 
(most frequency rows per time scattering rows), then halved with increasing `j2`.


This is frankly cryptic. Nothing in the paper definition of JTFS says that the shape is different by default. The user would have to go really into the weeds ot the discretization scheme to understand that.

"earliest coefficients then halved with increasing `j2`" is not the sort of sentence that a newcomer could understand.

**Reply by lostanlen** [reply]
The problem of doing this is not only that it's complicated but primarily that we have no experimental evidence thus far that it's useful, whereas for the basic implementation (#2 and #3) we can connect to a long slew of research on STRF + we can rely on our papers.

Your piece of doc does not answer the question of "why would i need it?"
and if your answer is "to make is the same size" then that could perfectly appply to the aligned implementation as well (#3)
so documenting it properly would require paragraphs of minute explanation without experimental grounding

**Reply by lostanlen** [reply]
after an offline conversation with @OverLordGoldDragon , here's the deal:

by default, out_type="array"  implements #3. out_type="array"  with some non-default kwarg implements #1.
 out_type="list"  implements #2

**Reply by OverLordGoldDragon** [reply]
Noting, with @lostanlen's agreement, #3's `subsample maximally` isn't valid and will alias, need `subsample minimally` just like #2.

**Comment by lostanlen** [comment, marked as answer]
@OverLordGoldDragon perhaps it's time to close this in light of https://github.com/lostanlen/kymatio/pull/1 and recent offline discussions?

**Reply by OverLordGoldDragon** [reply]
Just mention the agreement is 1, 2, and 3 as below in one of the answers for visibility, I'll close

```
  1. Pad varyingly,     subsample varyingly -> same out shape, unaligned
  2. Pad varyingly,     subsample minimally -> var  out shape, !aligned
  3. Pad same (to max), subsample minimally -> same out shape, aligned

pad varyingly = minimally per each `psi2`, 
subsample varyingly = proportionally to pad
subsample minimally = by largest subsampling factor in smallest padding
```

**Comment by OverLordGoldDragon** [comment]
Right-padding is the only padding supporting exact alignment. Suppose `x` is padded portion, and we progressively subsample:

```python
[x, x, x, x, 0, 1, 2, 3, 4, 5, 6, 7, x, x, x, x]
->
[x, x, 0, 2, 4, ...]
->
[x, 0, ...]
->
```

Compare with padding by two more on the left:

```python
[x, x, x, x, x, x, 0, 1, 2, 3, 4, 5, 6, 7, x, x, x, x]
->
[x, x, x, 0, 2, 4, ...]
->
[x, x, 2, ...]
```

After unpadding, first case ends up with `[0, ...]`, and second with `[2, ...]`. It's possible to guarantee `[0, ...]` for any subsampling, but not by padding to a dyadic length. In contrast, right padding's first sample will always end up at `0`.

However, left-right padding is preferred for performance: before `* phi_fr` we can unpad `* psi_fr` (case `J_fr > log2_F`) by slicing the array, obtaining a view. With right-padding we must circular-slice equivalently, which involves concatenating non-contiguous views: this makes a copy.

Prioritizing stride alignment is questionable; we already lose exact alignment if there's more than one signal component (e.g. AM tone), and additionally by not zero-padding. Perhaps the cost of the copy can be diminished via pre-allocation. I've yet to bench whether this unpadding grants a worthwhile speedup to begin with.

**Reply by OverLordGoldDragon** [reply]
Made a mistake: `True` does not contain more information, since `False` subsamples without aliasing. This makes `False` much more information-dense.
